src/owned.rs

```diff
@@ -0,0 +1,392 @@
+///         if *self.counter.borrow() % 2 == 0 {
+///             &[0x00]
+///         } else {
+///             &[0x01]
+///         }
+///     }
+///
+/// }
+/// ```
+/// The above code does not meet the safety contract because
+/// every other access will return a different set of bytes and it is not
+/// referencing a constant buffer.
+///
+/// Another example for possible unsafety is as follows:
+/// ```
+/// use rkyv_util::owned::StableBytes;
+/// 
+///
+/// struct Good {
+///     data: Vec<u8>
+/// }
+///
+/// unsafe impl StableBytes for Good {
+///     fn bytes(&self) -> &[u8] {
+///         self.data.as_ref()
+///     }
+/// }
+/// ```
+/// The above implementation is only safe if the `Vec<u8>` inside
+/// of `Good` is only ever mutated through `StableBytesMut`.
+/// Otherwise the bytes could be changed.
+///
+/// An example that is always safe would be the following:
+/// ```
+/// use rkyv_util::owned::StableBytes;
+/// use std::sync::Arc;
+/// 
+///
+/// struct Good {
+///     data: Arc<[u8]>
+/// }
+///
+/// unsafe impl StableBytes for Good {
+///     fn bytes(&self) -> &[u8] {
+///         self.data.as_ref()
+///     }
+/// }
+/// ```
+/// Since the data within the `Arc<[u8]>` is not mutable, we can never change
+/// it and thus this is necessarily stable.
+pub unsafe trait StableBytes {
+    /// Gets the underlying bytes.
+    fn bytes(&self) -> &[u8];
+}
+
+
+/// A contract guaranteeing that bytes should originate
+/// from the same source between acceses.
+///
+/// # Safety
+/// - `bytes` always refers to the same buffer.
+/// - The **only** way to edit the underlying bytes is through `StableBytesMut`
+///
+/// For instance, the following malicious implementation would be unsafe:
+/// ```
+/// use rkyv_util::owned::StableBytesMut;
+/// use std::cell::RefCell;
+///
+/// struct Malicious {
+///     counter: RefCell<u8>,
+///     buf_one: Vec<u8>,
+///     buf_two: Vec<u8>
+/// }
+///
+/// unsafe impl StableBytesMut for Malicious {
+///     fn bytes_mut(&mut self) -> &mut [u8] {
+///         *self.counter.borrow_mut() += 1;
+///         if *self.counter.borrow() % 2 == 0 {
+///             &mut self.buf_one
+///         } else {
+///             &mut self.buf_two
+///         }
+///     }
+///
+/// }
+/// ```
+/// The above code does not meet the safety contract because
+/// every other access will return a different set of bytes and it is not
+/// referencing a constant buffer.
+///
+/// Another example for possible unsafety is as follows:
+/// ```
+/// use rkyv_util::owned::StableBytesMut;
+/// 
+///
+/// struct Good {
+///     data: Vec<u8>
+/// }
+///
+/// unsafe impl StableBytesMut for Good {
+///     fn bytes_mut(&mut self) -> &mut [u8] {
+///         self.data.as_mut()
+///     }
+/// }
+/// ```
+/// The above implementation is only safe if the `Vec<u8>` inside
+/// of `Good` is only ever mutated through `StableBytesMut`.
+/// Otherwise the bytes could be changed.
+pub unsafe trait StableBytesMut {
```

Comment:
```suggestion
pub unsafe trait StableBytesMut: StableBytes {
```
This is just an ergonomics improvement so that if someone wants to use a container mutably and immutably, they can just bound `C: StableBytesMut`.